### PR TITLE
Sheet: new onAnimationEnd prop + fix for bug

### PR DESCRIPTION
**Pull request by AlbertCarreras**
### Summary

This PR adds onAnimationEnd to Sheet so events can get triggered once the component is fully rendered or removed. Examples can be placing educational focus on elements inside Sheet.

To detect which animation is (in vs out) the callback `({ animationState }) => void` has access to `animationState`

Kapture 2021-08-12 at 19 57 13

ALSO, fix for [BUG-131248 Sheet dismiss button unreachable when heading is missing
](https://jira.pinadmin.com/browse/BUG-131248)

### Links

- [Jira](link to Jira ticket(s))
- [TDD](link to Paper doc)
- [Figma](link to Figma file)

### Checklist

- [ ] Added unit and Flow Tests
- [ ] Added documentation + accessibility tests
- [ ] Verified accessibility: keyboard & screen reader interaction
- [ ] Checked dark mode, responsiveness, and right-to-left support
- [ ] Checked stakeholder feedback (e.g. Gestalt designers)

### CODE REVIEW

File: docs/src/Sheet.doc.js [hunk 1]
```diff
@@ -1033,6 +1039,8 @@ function ClosingExample(props) {
       - Done red button (children)
       - Left arrow red icon button (children)
       - Close button (footer)
+
+      Sheet also provides \`onAnimationEnd\`, a callback that gets triggered at the end of each animation. The callback has access to \`animationState\` to identify the end of each 'in' and 'out' animation for cases where each event don't trigger the same response.
```
Comment (dangerismycat) [inline docs/src/Sheet.doc.js:1043]: grammar nit: "where each event doesn't trigger the same response"

Comment (ayeshakmaz) [inline docs/src/Sheet.doc.js:1043]: Maybe: "Sheet also provides \`onAnimationEnd\`, a callback that gets triggered at the end of each animation. The callback has access to \`animationState\` to identify the end of each 'in' and 'out' animation for cases where the two events trigger different responses."

File: packages/gestalt/src/Sheet.js [hunk 2]
```diff
@@ -117,14 +121,18 @@ function Sheet(props: SheetProps): Node {
...
     onDismiss,
     size = 'sm',
     subHeading,
   } = props;
 
   const [showTopShadow, setShowTopShadow] = useState<boolean>(false);
   const [showBottomShadow, setShowBottomShadow] = useState<boolean>(false);
-  const { animationState, onAnimationEnd } = useAnimation();
+  const {
+    animationState: animationStateHook,
```
Comment (dangerismycat) [inline packages/gestalt/src/Sheet.js:133]: nit: this name is a bit odd, since it makes it sound like this variable is a hook itself. Perhaps stick with the convention you used for the next one, `animationStateFromHook`?

**Review by dangerismycat — APPROVED** [review]

File: packages/gestalt/src/Sheet.js [hunk 3]
```diff
@@ -142,6 +150,11 @@ function Sheet(props: SheetProps): Node {
     };
   }, [onDismiss]);
 
+  const handleOnAnimationEnd = useCallback(() => {
+    onAnimationEndFromHook?.();
+    onAnimationEnd?.({ animationState: animationStateHook === 'in' ? 'in' : 'out' });
```
Comment (ayeshakmaz) [inline packages/gestalt/src/Sheet.js:155]: Can this not be simplified to "animationState: animationStateHook"?

File: packages/gestalt/src/Sheet.js [hunk 4]
```diff
@@ -213,7 +227,13 @@ function Sheet(props: SheetProps): Node {
                 )}
                 {!heading && (
                   <Box display="flex" flex="grow" justifyContent="end" marginBottom={8}>
-                    <Box flex="none" paddingX={6} paddingY={7} position="absolute">
+                    <Box
+                      flex="none"
+                      paddingX={6}
+                      paddingY={7}
+                      position="absolute"
+                      zIndex={new FixedZIndex(1)}
```
Comment (AlbertCarreras) [inline packages/gestalt/src/Sheet.js:235]: FIX for bug https://jira.pinadmin.com/browse/BUG-131248

File: docs/src/Sheet.doc.js (hunk 1 above)
Comment (AlbertCarreras) [reply docs/src/Sheet.doc.js:1043]: DONE

File: packages/gestalt/src/Sheet.js (hunk 2 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/Sheet.js:134]: OOH I didn't notice that subtlety, yeah 
DONE

File: packages/gestalt/src/Sheet.js (hunk 3 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/Sheet.js:156]: @ayeshakmaz I didn't understand the question, can you rephrase it?

Comment (AlbertCarreras) [reply packages/gestalt/src/Sheet.js]: OOH! `animationStateHook` returns only 'in' & 'out' but the type is `export type AnimationStateType = 'in' | 'postIn' | 'out' | 'postOut' | null;` We wanna simplify the type to just   `'in' | 'out'` and we just pass `animationStateHook` we'd have to extend its type. We can only access 'in' & 'out' at the end of the animation, but the type doesn't match with the behavior in this particular moment.
